crates/wasmtime/src/compiler.rs

```diff
@@ -0,0 +1,700 @@
+        Self {
+            namespace: Self::WASM_TO_NATIVE_TRAMPOLINE_KIND,
+            index: index.as_u32(),
+        }
+    }
+}
+
+#[cfg(feature = "component-model")]
+impl CompileKey {
+    const LOWERING_KIND: u32 = 0b100 << 29;
+    const ALWAYS_TRAP_KIND: u32 = 0b101 << 29;
+    const TRANSCODER_KIND: u32 = 0b110 << 29;
+
+    fn lowering(index: wasmtime_environ::component::LoweredIndex) -> Self {
+        Self {
+            namespace: Self::LOWERING_KIND,
+            index: index.as_u32(),
+        }
+    }
+
+    fn always_trap(index: wasmtime_environ::component::RuntimeAlwaysTrapIndex) -> Self {
+        Self {
+            namespace: Self::ALWAYS_TRAP_KIND,
+            index: index.as_u32(),
+        }
+    }
+
+    fn transcoder(index: wasmtime_environ::component::RuntimeTranscoderIndex) -> Self {
+        Self {
+            namespace: Self::TRANSCODER_KIND,
+            index: index.as_u32(),
+        }
+    }
+}
+
+#[derive(Clone, Copy)]
+enum CompiledFunction<T> {
+    Function(T),
+    #[cfg(feature = "component-model")]
+    AllCallFunc(wasmtime_environ::component::AllCallFunc<T>),
+}
+
+impl<T> CompiledFunction<T> {
+    fn unwrap_function(self) -> T {
+        match self {
+            Self::Function(f) => f,
+            #[cfg(feature = "component-model")]
+            Self::AllCallFunc(_) => panic!("CompiledFunction::unwrap_function"),
+        }
+    }
+
+    #[cfg(feature = "component-model")]
+    fn unwrap_all_call_func(self) -> wasmtime_environ::component::AllCallFunc<T> {
+        match self {
+            Self::AllCallFunc(f) => f,
+            Self::Function(_) => panic!("CompiledFunction::unwrap_all_call_func"),
+        }
+    }
+}
+
+#[cfg(feature = "component-model")]
+impl<T> From<wasmtime_environ::component::AllCallFunc<T>> for CompiledFunction<T> {
+    fn from(f: wasmtime_environ::component::AllCallFunc<T>) -> Self {
+        Self::AllCallFunc(f)
+    }
+}
+
+struct CompileOutput {
+    key: CompileKey,
+    symbol: String,
+    function: CompiledFunction<Box<dyn Any + Send>>,
+    info: Option<WasmFunctionInfo>,
+}
+
+/// The collection of things we need to compile for a Wasm module or component.
+pub struct CompileInputs<'a> {
+    inputs: Vec<CompileInput<'a>>,
+}
+
+fn push_input<'a, 'data>(
+    inputs: &mut Vec<CompileInput<'a>>,
+    f: impl FnOnce(&Tunables, &dyn Compiler) -> Result<CompileOutput> + Send + 'a,
+) {
+    inputs.push(Box::new(f) as _);
+}
+
+impl<'a> CompileInputs<'a> {
+    /// Create the `CompileInputs` for a core Wasm module.
+    pub fn for_module(
+        types: &'a ModuleTypes,
+        translation: &'a ModuleTranslation<'a>,
+        functions: PrimaryMap<DefinedFuncIndex, FunctionBodyData<'a>>,
+    ) -> Self {
+        let mut inputs = vec![];
+        let module_index = StaticModuleIndex::from_u32(0);
+
+        Self::collect_inputs_in_translations(
+            types,
+            std::iter::once((module_index, translation, functions)),
```

Comment:
No changes necessary, but I've found that `[expr]` works nowadays for "iterate this thing once" which I think reads pretty well too

Reply:
Oh nice, I always forget that arrays are useful now.